src/sensai/featuregen.py

```diff
@@ -195,44 +221,62 @@ def _generateFeatureDict(self, namedTuple) -> Dict[str, Any]:
 
 
 class FeatureGeneratorTakeColumns(RuleBasedFeatureGenerator):
-    def __init__(self, columns: Union[str, List[str]], categoricalFeatureNames: Sequence[str] = (),
-            normalisationRules: Sequence[data_transformation.DFTNormalisation.Rule] = ()):
-        super().__init__(categoricalFeatureNames=categoricalFeatureNames, normalisationRules=normalisationRules)
+    def __init__(self, columns: Union[str, List[str]] = None, exceptColumns: List[str] = None, categoricalFeatureNames: Sequence[str] = (),
+                 normalisationRules: Sequence[data_transformation.DFTNormalisation.Rule] = (),
+                 normalisationRuleTemplate: data_transformation.DFTNormalisation.RuleTemplate = None):
+        """
+
+        :param columns: name of the column or list of names of columns to be taken. If None, all columns will be taken.
+        :param exceptColumns: list of names of columns to not take if present in the input df
+        :param categoricalFeatureNames:
+        :param normalisationRules:
+        :param normalisationRuleTemplate:
+        """
+        super().__init__(categoricalFeatureNames=categoricalFeatureNames, normalisationRules=normalisationRules, normalisationRuleTemplate=normalisationRuleTemplate)
         if isinstance(columns, str):
             columns = [columns]
         self.columns = columns
+        self.exceptColumns = exceptColumns if exceptColumns is not None else []
 
     def _generate(self, df: pd.DataFrame, ctx=None) -> pd.DataFrame:
-        missingCols = set(self.columns) - set(df.columns)
+        columnsToTake = self.columns if self.columns is not None else df.columns
+        columnsToTake = set(columnsToTake).difference(self.exceptColumns)
+
+        missingCols = columnsToTake.difference(df.columns)
         if len(missingCols) > 0:
             raise Exception(f"Columns {missingCols} not present in data frame; available columns: {list(df.columns)}")
-        return df[self.columns]
-
 
-class FeatureGeneratorTakeAllColumns(RuleBasedFeatureGenerator):
-    def _generate(self, df: pd.DataFrame, ctx=None) -> pd.DataFrame:
-        return df
+        return df[sorted(columnsToTake)]
```

Comment:
Typically, column order should not matter, but maybe it would be better if we didn't override the user's ordering given in self.columns (for the case where he specifies it). We can filter out exceptColumns with a list comprehension instead of using sets.

Reply:
Done